test/unit/rayfireAMR_test.C

```diff
@@ -154,6 +156,143 @@ namespace GRINSTesting
+      CPPUNIT_ASSERT( !(rayfire->map_to_rayfire_elem(5)) );
+      CPPUNIT_ASSERT( !(rayfire->map_to_rayfire_elem(1)) );
+      CPPUNIT_ASSERT( !(rayfire->map_to_rayfire_elem(7)) );
+
+      // now do a uniform refinement
+      libMesh::MeshRefinement mr(*mesh);
+      mr.uniformly_refine();
+
+      rayfire->reinit(*mesh);
+
+      // check post-refinement
+      CPPUNIT_ASSERT( rayfire->map_to_rayfire_elem(mesh->elem(0)->child(0)->id()) );
+      CPPUNIT_ASSERT( rayfire->map_to_rayfire_elem(mesh->elem(0)->child(1)->id()) );
+      CPPUNIT_ASSERT( rayfire->map_to_rayfire_elem(mesh->elem(6)->child(0)->id()) );
+      CPPUNIT_ASSERT( rayfire->map_to_rayfire_elem(mesh->elem(9)->child(1)->id()) );
+      CPPUNIT_ASSERT( rayfire->map_to_rayfire_elem(mesh->elem(8)->child(2)->id()) );
+
+      // coarsen two of the TRIs
+      for (unsigned int c=0; c<mesh->elem(6)->n_children(); c++)
+        {
+          mesh->elem(6)->child(c)->set_refinement_flag(libMesh::Elem::RefinementState::COARSEN);
+          mesh->elem(9)->child(c)->set_refinement_flag(libMesh::Elem::RefinementState::COARSEN);
+        }
+
+      mr.coarsen_elements();
+      rayfire->reinit(*mesh);
+
+      CPPUNIT_ASSERT( mesh->elem(6)->active() );
+      CPPUNIT_ASSERT( mesh->elem(9)->active() );
+
+      // check post-coarsen
+      CPPUNIT_ASSERT( rayfire->map_to_rayfire_elem(mesh->elem(0)->child(0)->id()) );
+      CPPUNIT_ASSERT( rayfire->map_to_rayfire_elem(mesh->elem(0)->child(1)->id()) );
+      CPPUNIT_ASSERT( rayfire->map_to_rayfire_elem(mesh->elem(8)->child(2)->id()) );
+      CPPUNIT_ASSERT( rayfire->map_to_rayfire_elem(6) );
+      CPPUNIT_ASSERT( rayfire->map_to_rayfire_elem(9) );
+
+      for (unsigned int c=0; c<mesh->elem(6)->n_children(); c++)
+        {
+          CPPUNIT_ASSERT( !(rayfire->map_to_rayfire_elem(mesh->elem(6)->child(c)->id())) );
+          CPPUNIT_ASSERT( !(rayfire->map_to_rayfire_elem(mesh->elem(9)->child(c)->id())) );
+        }
+    }
+
+    //! Mesh given to init() is already refined
+    void start_with_refined_mesh()
+    {
+      GRINS::SharedPtr<libMesh::UnstructuredMesh> mesh = build_quad4_elem();
+
+      // uniform refinement
+      libMesh::MeshRefinement mr(*mesh);
+      mr.uniformly_refine();
+
+      // now init the rayfire
+      libMesh::Point origin(0.0,0.1);
+      libMesh::Point end_point(1.0,0.1);
+      libMesh::Real theta = calc_theta(origin,end_point);
+
+      GRINS::SharedPtr<GRINS::RayfireMesh> rayfire = new GRINS::RayfireMesh(origin,theta);
+      rayfire->init(*mesh);
+
+      CPPUNIT_ASSERT( rayfire->map_to_rayfire_elem(mesh->elem(0)->child(0)->id()) );
+      CPPUNIT_ASSERT( rayfire->map_to_rayfire_elem(mesh->elem(0)->child(1)->id()) );
+
+      CPPUNIT_ASSERT( !(rayfire->map_to_rayfire_elem(mesh->elem(0)->child(2)->id())) );
+      CPPUNIT_ASSERT( !(rayfire->map_to_rayfire_elem(mesh->elem(0)->child(3)->id())) );
+
+      // refine again
+      mr.uniformly_refine();
```

Comment:
Following this block of checks, can you uniformly_coarsen(); rayfire->reinit; uniformly_coarsen; rayfire->reinit and make sure we're doing the right thing?